src/bootstrap/builder.rs

```diff
@@ -302,8 +302,10 @@ impl StepDescription {
         }
     }
 
-    fn maybe_run(&self, builder: &Builder<'_>, pathsets: Vec<PathSet>) {
-        if pathsets.iter().any(|set| self.is_excluded(builder, set)) {
+    fn maybe_run(&self, builder: &Builder<'_>, mut pathsets: Vec<PathSet>) {
+        pathsets.retain(|set| !self.is_excluded(builder, set));
+
+        if pathsets.is_empty() || pathsets.len() == 0 {
```

Comment:
This is redundant, `is_empty()` already checks for `len() == 0`

```suggestion
        if pathsets.is_empty() {
```